opal/mca/accelerator/accelerator.h

```diff
@@ -0,0 +1,475 @@
+ * can specify the transfer type to avoid pointer detection for
+ * performance.
+ *
+ * @param[IN] dev_id         Associated device to copy to/from
+ * @param[IN] dest           Destination to copy memory to
+ * @param[IN] dpitch         Pitch of destination memory
+ * @param[IN] src            Source to copy memory from
+ * @param[IN] spitch         Pitch of source memory
+ * @param[IN] width          Width of matrix transfer (columns in bytes)
+ * @param[IN] height         Height of matrix transfer (rows)
+ * @param[IN] type           Transfer type field for performance
+ *                           Can be set to MCA_ACCELERATOR_TRANSFER_UNSPEC
+ *                           if caller is unsure of the transfer direction.
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_matrix_memcpy_fn_t)(
+    int dev_id, void *dest, size_t dpitch, const void *src, size_t spitch,
+    size_t width, size_t height, opal_accelerator_transfer_type_t type);
+
+/**
+ * Copies memory synchronously from src to dest. Memory of dest and src
+ * may overlap. Optionally can specify the transfer type to
+ * avoid pointer detection for performance.
+ *
+ * @param[IN] dev_id         Associated device to copy to/from
+ * @param[IN] dest           Destination to copy memory to
+ * @param[IN] src            Source to copy memory from
+ * @param[IN] size           Size of memory to copy
+ * @param[IN] type           Transfer type field for performance
+ *                           Can be set to MCA_ACCELERATOR_TRANSFER_UNSPEC
+ *                           if caller is unsure of the transfer direction.
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_memmove_fn_t)(
+    int dev_id, void *dest, const void *src, size_t size,
+    opal_accelerator_transfer_type_t type);
+
+/**
+ * Allocates size bytes memory from the device and sets ptr to the
+ * pointer of the allocated memory. The memory is not initialized.
+ *
+ * @param[IN] dev_id         The device to allocate memory on
+ * @param[OUT] ptr           Returns pointer to allocated memory
+ * @param[IN] size           Size of memory to allocate
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_malloc_fn_t)(
+    int dev_id, void **ptr, size_t size);
+
+/**
+ * Frees the memory space pointed to by ptr which has been returned by
+ * a previous call to an opal_accelerator_base_module_malloc_fn_t().
+ * If the function is called on a ptr that has already been freed,
+ * undefined behavior occurs. If ptr is NULL, no operation is performed,
+ * and the function returns OPAL_SUCCESS.
+ *
+ * @param[IN] ptr            Pointer to free
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ */
+typedef int (*opal_accelerator_base_module_free_fn_t)(
```

Comment:
habana needs a device id to free device buffer.

Reply:
Unfortunately there is not such an API in habana yet

Habana does not have an exposed API to query which device is being used in current process, or which device is on for a given device buffer.

Application needs to acquire a device with device id explicitly, and pass the device id to open mpi.

Reply:
We cannot touch the MPI application API because it takes years to get changes in (with the current MPI API, it is impossible for the user to pass dev_id through the interface).  Under the current design constraints, the Habana accelerator component would need some global hack (similar to a CUDA context) so that it can fetch the dev_id it needs.

It is unfortunate that this "hack"/"abstraction failure" ties Applications to specific accelerator hardware.

This hack could be as simple as:
```
/* Habana app.c */
int habana_dev_id = 0;

/* Habana Accelerator Component */
extern int habana_dev_id;
```

Reply:
I'd strongly recommend environment variables (or, better yet, MCA parameters!) for this, not an extern.  But, again, uplevel the conversation to the framework's expected use behaviors.

Reply:
how about an info object (MPI_Comm_set_info, MPI_Win_set_info/Win_create)?

Reply:
If you remove the device id from the API and every component has to handle it themselves, then your invariant is that there is exactly one device per process.

Reply:
Just as a note, we're not actually thinking about adding Habana support right now, I was just trying to see how generic our API was, since the current cuda implementation doesn't rely on device ids. I'm starting to lean on the boat of getting rid for device ids entirely and letting the implementations handle it, but setting info objects could be a solution.